sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.implementation.caches;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+/**
+ *
+ * @param <TKey>
+ * @param <TValue>
+ *
+ * This is a thread safe AsyncCache that allows refreshing values in the background.
+ * The benefits of AsyncCacheNonBlocking over AsyncCache is it keeps stale values until the refresh is completed.
+ * AsyncCache removes values causing it to block all requests until the refresh is complete.
+ * 1. For example 1 replica moved out of the 4 replicas available. 3 replicas could still be processing requests.
+ *    The request going to the 1 stale replica would be retried.
+ * 2. AsyncCacheNonBlocking updates the value in the cache rather than recreating it on each refresh. This will help reduce object creation.
+ *
+ */
+public class AsyncCacheNonBlocking<TKey, TValue> {
+    private final static Logger logger = LoggerFactory.getLogger(AsyncCacheNonBlocking.class);
+    private final ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values;
+    private final IEqualityComparer<TKey> keyEqualityComparer;
+
+    public AsyncCacheNonBlocking(ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values, IEqualityComparer<TKey> keyEqualityComparer) {
```

Comment:
maybe change from public to private since it is not being used anywhere else?

Reply:
Fixed

Reply:
it is still showing as public for me ?

Reply:
This constructor AsyncCacheNonBlocking() has to be public as the RxPartitionCache uses it. I have made the other two private.

Reply:
Nm, I started too quick before you have pushed all your changes, so all good here